### PR TITLE
Add support for colored box shadows

### PR DESCRIPTION
**Pull request by adamwathan**
This PR adds support for specifying a box shadow color using a composable API:

```html
<div class="shadow-xl shadow-blue-500/20">
```

The motivation for this is to make it possible to tint your shadows to look better on colored backgrounds, which is a design principle that @joshwcomeau outlines beautifully in this blog post:

https://www.joshwcomeau.com/css/designing-shadows/

Here you can see the default black-based shadow on the left, and a tinted shadow on the right:

image

The difference is kind of subtle but the tinted ones look more natural when you get the values right.

The composable API is very flexible (almost too flexible really), so this is sort of an advanced feature that non-designers might not have an easy time getting great results with. I hope to provide some guidance around how to approach using these APIs in the documentation for v3 when it's released.

Under the hood, **all colors in a shadow are replaced by the color you specify**. For example, given this shadow in your config:

```js
boxShadow: {
  // ...
  fancy: '0 1px 3px rgba(24, 53, 76, 0.11), 0 2px 9px rgba(11, 22, 99, 0.42)'
}
```

...if you write the following HTML:

```html
<div class="shadow-fancy shadow-blue-500">
```

...the final rendered box shadow will be this:

```
box-shadow: 0 1px 3px #3B82F6, 0 2px 9px #3B82F6
```

Notice that even though in the original shadow the two separate shadow layers had separate colors, in the final shadow **both layers are the exact same color**. This is the only practical approach I felt was reasonable after attempting to approach this feature from scratch like 12 times.

This is a little weird because the default Tailwind shadow use different alpha values for the shadow layers. For this reason, **this PR also tweaks the default shadows to use the same alpha value for each layer**. By doing this, when you specify a shadow color the colored shadow will have the same transparency and blending characteristics as the default shadow.

We've tried to make the new shadows as visually identical to the old ones as possible, which you can preview in this Play:

https://play.tailwindcss.com/S8d4uoZ2sz?layout=preview

It's also worth noting that if you change a shadow at a different breakpoint, **it will override any custom color**.

```html
<!-- This shadow will NOT be red at md screens and above -->
<div class="shadow-lg shadow-red-500 md:shadow-xl">
```

To preserve the color, it needs to be respecified:

```html
<!-- This shadow WILL be red at md screens and above -->
<div class="shadow-lg shadow-red-500 md:shadow-xl md:shadow-red-500">
```

This is the only way we could make it possible to switch from a color back to the default shadow color at a different breakpoint. In the future we may come up with a terser syntax (like `shadow-xl/red-500`) to make this feel less annoying, but in practice I don't expect this to be a huge pain point anyways.

All of the colors are available as shadow colors by default, and if you need to customize them separately you can do so under the `boxShadowColor` key in your config.

### CODE REVIEW

**Comment by Soviut** [comment]
I would love if it were possible to control the shadow alpha, with either `shadow-alpha-50` or `shadow-red-500/50`.

**Comment by adamwathan** [comment]
> I would love if it were possible to control the shadow alpha, with either `shadow-alpha-50` or `shadow-red-500/50`.

Yeah this works as part of this PR too 👍🏻

`shadow-red-500/20`

File: src/corePlugins.js [hunk 1]
```diff
@@ -1782,18 +1784,42 @@ export let corePlugins = {
           shadow: (value) => {
             value = transformValue(value)
 
+            let ast = parseBoxShadowValue(value)
+            for (let shadow of ast) {
+              // Don't override color if the whole shadow is a variable
+              if (shadow.x === undefined || shadow.y === undefined) {
+                continue
+              }
```
Comment (RobinMalfait) [inline src/corePlugins.js:1792]: Should we encode this information in each shadow itself? Like a `isValid` flag or something?
Not a requirement, but I noticed the same logic in multiple places.

**Review by RobinMalfait — APPROVED** [review]
Beautiful!

File: src/corePlugins.js (hunk 1 above)
Comment (adamwathan) [reply src/corePlugins.js:1792]: Yeah I agree or even have the parse function return an error or something if it doesn’t parse a proper value. Moar tuples? Is this Elm?

**Comment by robdekort** [comment]
This is beautiful and still so flexible. Lovely!

**Comment by snaptopixel** [comment]
Seems like a nice hands-off approach could be to calculate and include a shadow-color variable as part of the `bg-` classes and have the `shadow-` utils use it by default? Something like https://codepen.io/fathom-casey/pen/VwzQydd

**Comment by Soviut** [comment]
@snaptopixel I feel like that makes a lot of assumptions about elements being nested and would result in a lot of confusion as to why shadows are behaving differently when the user hasn't explicitly applied classes to them.

An example where I use shadows without nesting are with modals that have an overlay that darkens the background. The overlay is a separate element while the floating modal dialog is a sibling, rather than a descendant. The background colour of the overlay wouldn't transmit to the dialog so I'd have to be manually setting shadow colour in that case.

**Comment by benjamincrozat** [comment]
Will this be available in version 2 since it's not a breaking change?

**Comment by noam-elbaz** [comment]
Can't contribute anything intelligent to the development of this feature other than - awesome! looking forward.  this is useful.

**Comment by captlevi01** [comment]
on which version can we try this?

**Pull request merged**